File: python/problem1/modelFitVal_solution.py
from pylab import *
from numpy.random import shuffle
from sklearn.linear_model import LogisticRegression


def modelFitVal(X, L, kPart):
    """
    modelFitVal takes multivariate data (one feature per column) and
    associated labels and returns the cross-validated performance of a linear
    model.

    """

    # set up shuffled data partitions
    nSamples = len(X)
    partTag = arange(nSamples) % kPart
    shuffle(partTag)

    pCorrect = 0

    for iPart in range(kPart):

        # separate training and test data
        index = (partTag == iPart)
        xTest = X[index]
        lTest = L[index]
        xTrain = X[~index]
        lTrain = L[~index]
        
        # fit model to training data using logistic regressio
        model = LogisticRegression()
        model.fit(xTrain, lTrain)

        # compute p(C1|D) for test data
        posterior_C1 = model.predict(xTest)

        # compare model's classification to actual labels
        pCorrect += mean(posterior_C1 == lTest) / kPart

    return pCorrect

File: matlab/problem1/modelFitVal_solution.m
function pCorrect = modelFitVal_solution(X, L, k)

% modelFitVal takes multivariate data (one feature per column) and
% associated labels and returns the cross-validated performance of a linear
% model.

% create a random permutation of sample indices
nSamples = size(X, 1);
randIdx = randperm(nSamples);

pCorrect = 0;

for iPart=1:k
    % separate training and test data
    idxTest = randIdx(iPart:k:nSamples);
    idxTrain = randIdx(setdiff(1:nSamples, idxTest));
    xTest = X(idxTest, :);
    xTrain = X(idxTrain, :);
    lTest = L(idxTest);
    lTrain = L(idxTrain);
    % fit logit model to training data
    coeff = glmfit(xTrain, lTrain, 'binomial', 'link', 'logit');
    % compute labels for test data from rounded p(C1|D)
    lPredicted = round(glmval(coeff, xTest, 'logit'));
    % compare predicted to actual labels
    pCorrect = pCorrect + 1/k*mean(lPredicted==lTest);
end
